### PR TITLE
rgwlc: dispatch S3 notifications on transition and mpu abort

### PR DESCRIPTION
**Pull request by mattbenjamin**
Fixes: https://tracker.ceph.com/issues/58641

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by dang — APPROVED** [review]

File: src/rgw/rgw_lc.cc [hunk 1]
```diff
@@ -828,15 +828,38 @@ int RGWLC::handle_multipart_expiration(rgw::sal::Bucket* target,
...
+	= target->get_object(key);
+
+      std::unique_ptr<rgw::sal::Notification> notify
+	= driver->get_notification(
+	  this, sal_obj.get(), nullptr, event_type,
+	  target, lc_id,
+	  const_cast<std::string&>(target->get_tenant()),
+	  lc_req_id, null_yield);
+
+      ret = notify->publish_reserve(this, nullptr);
```
Comment (yuvalif) [inline src/rgw/rgw_lc.cc:851]: `publish_reserve()` may fail in the case of persistent notifications. we should check `ret`

File: src/rgw/rgw_lc.cc [hunk 2]
```diff
@@ -828,15 +828,38 @@ int RGWLC::handle_multipart_expiration(rgw::sal::Bucket* target,
...
+
       int ret = mpu->abort(this, cct);
       if (ret == 0) {
-        if (perfcounter) {
+
+        (void) notify->publish_commit(
```
Comment (yuvalif) [inline src/rgw/rgw_lc.cc:856]: `publish_commit()` may also fail. although there is nothing much to do except logging an error message

File: src/rgw/rgw_lc.cc [hunk 3]
```diff
@@ -1661,6 +1735,15 @@ int RGWLC::bucket_lc_process(string& shard_id, LCWorker* worker,
     worker->workpool->drain();
   }
 
+  std::unique_ptr<rgw::sal::User> user;
```
Comment (yuvalif) [inline src/rgw/rgw_lc.cc:1738]: how is this change related to notifications?

**Comment by yuvalif** [comment]
could you please update documentation: https://docs.ceph.com/en/latest/radosgw/s3-notification-compatibility/
(currently it says: "Defined, Ceph extension (not generated)")

File: src/rgw/rgw_lc.cc [hunk 4]
```diff
@@ -1268,27 +1291,78 @@ class LCOpAction_Transition : public LCOpAction {
...
-    int ret = oc.obj->transition_to_cloud(oc.bucket, oc.tier.get(), oc.o,
-					  oc.env.worker->get_cloud_targets(), oc.cct,
-					  !delete_object, oc.dpp, null_yield);
+    /* notifications */
+    std::unique_ptr<rgw::sal::Bucket> bucket;
+    std::unique_ptr<rgw::sal::Object> obj;
+    auto& bucket_info = oc.bucket->get_info();
+    std::string version_id;
+
+    ret = oc.driver->get_bucket(nullptr, bucket_info, &bucket);
```
Comment (yuvalif) [inline src/rgw/rgw_lc.cc:1315]: what is the difference between `bucket` and `oc.bucket`?

Comment (yuvalif) [reply src/rgw/rgw_lc.cc:1315]: similar questions on `user` and `obj`?

File: src/rgw/rgw_lc.cc [hunk 5]
```diff
@@ -828,15 +828,38 @@ int RGWLC::handle_multipart_expiration(rgw::sal::Bucket* target,
   params.ns = RGW_OBJ_NS_MULTIPART;
   params.access_list_filter = &mp_filter;
 
+  auto event_type = rgw::notify::ObjectExpirationAbortMPU;
+  std::string version_id;
```
Comment (yuvalif) [inline src/rgw/rgw_lc.cc:832]: i dont see `version_id` being set anywhere. could just be an empty string passed to `publish_commit()`

File: src/rgw/rgw_lc.cc (hunk 3 above)
Comment (mattbenjamin) [reply src/rgw/rgw_lc.cc:1738]: it is, or was, required to fetch the correct state into the sal object

File: src/rgw/rgw_lc.cc (hunk 5 above)
Comment (mattbenjamin) [reply src/rgw/rgw_lc.cc:832]: (this matches the other usage in rgw_lc.cc, so don't want to change it)

File: src/rgw/rgw_lc.cc [hunk 6]
```diff
@@ -828,15 +828,38 @@ int RGWLC::handle_multipart_expiration(rgw::sal::Bucket* target,
   params.ns = RGW_OBJ_NS_MULTIPART;
   params.access_list_filter = &mp_filter;
 
+  auto event_type = rgw::notify::ObjectExpirationAbortMPU;
```
[the remark this answers is not in the document]
Comment (mattbenjamin) [reply src/rgw/rgw_lc.cc:831]: for clarity;  constifying it doesn't do much

File: src/rgw/rgw_lc.cc [hunk 7]
```diff
@@ -828,15 +828,38 @@ int RGWLC::handle_multipart_expiration(rgw::sal::Bucket* target,
...
       std::unique_ptr<rgw::sal::MultipartUpload> mpu = target->get_multipart_upload(key.name);
+
+      std::unique_ptr<rgw::sal::Object> sal_obj
+	= target->get_object(key);
+
+      std::unique_ptr<rgw::sal::Notification> notify
+	= driver->get_notification(
+	  this, sal_obj.get(), nullptr, event_type,
+	  target, lc_id,
+	  const_cast<std::string&>(target->get_tenant()),
```
[the remark this answers is not in the document]
Comment (mattbenjamin) [reply src/rgw/rgw_lc.cc:848]: legacy, I guess

File: src/rgw/rgw_lc.cc (hunk 1 above)
Comment (mattbenjamin) [reply src/rgw/rgw_lc.cc:851]: ok

File: src/rgw/rgw_lc.cc [hunk 8]
```diff
@@ -828,15 +828,38 @@ int RGWLC::handle_multipart_expiration(rgw::sal::Bucket* target,
...
+	  lc_req_id, null_yield);
+
+      ret = notify->publish_reserve(this, nullptr);
+
       int ret = mpu->abort(this, cct);
```
[the remark this answers is not in the document]
Comment (mattbenjamin) [reply src/rgw/rgw_lc.cc:853]: copy-paste, I guess

File: src/rgw/rgw_lc.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (mattbenjamin) [reply src/rgw/rgw_lc.cc:831]: (so not changing this)

File: src/rgw/rgw_lc.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (mattbenjamin) [reply src/rgw/rgw_lc.cc:848]: will fix

**Comment by mattbenjamin** [comment]
> could you please update documentation: https://docs.ceph.com/en/latest/radosgw/s3-notification-compatibility/ (currently it says: "Defined, Ceph extension (not generated)")

thanks, forgot about that

File: src/rgw/rgw_lc.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (mattbenjamin) [reply src/rgw/rgw_lc.cc:848]: /home/mbenjamin/ceph-cp/src/rgw/rgw_lc.cc: In lambda function:
/home/mbenjamin/ceph-cp/src/rgw/rgw_lc.cc:848:29: error: binding reference of type ??std::string&?? {aka ??std::__cxx11::basic_string<char>&??} to ??const std::string?? {aka ??const std::__cxx11::basic_string<char>??} discards qualifiers
  848 |           target->get_tenant(),
      |           ~~~~~~~~~~~~~~~~~~^~
In file included from /home/mbenjamin/ceph-cp/src/rgw/rgw_bucket.h:15,
                 from /home/mbenjamin/ceph-cp/src/rgw/rgw_lc.cc:26:
/home/mbenjamin/ceph-cp/src/rgw/rgw_sal.h:329:103: note:   initializing argument 7 of ??virtual std::unique_ptr<rgw::sal::Notification> rgw::sal::Driver::get_notification(const DoutPrefixProvider*, rgw::sal::Object*, rgw::sal::Object*, rgw::notify::EventType, rgw::sal::Bucket*, std::string&, std::string&, std::string&, optional_yield)??
  329 |     rgw::notify::EventType event_type, rgw::sal::Bucket* _bucket, std::string& _user_id, std::string& _user_tenant,
      |                                                                                          ~~~~~~~~~~~~~^~~~~~~~~~~~

File: src/rgw/rgw_lc.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (mattbenjamin) [reply src/rgw/rgw_lc.cc:853]: fixed

**Comment by mattbenjamin** [comment]
the version @yuvalif reviewed passed an rgw suite run:
https://pulpito.ceph.com/mbenjamin-2023-03-29_17:09:09-rgw-wip-notify-lctrans-distro-default-smithi/
(but does not include new functional tests)

File: src/rgw/rgw_lc.cc (hunk 2 above)
Comment (mattbenjamin) [reply src/rgw/rgw_lc.cc:863]: if you think logging an errmsg there is important, will change this

File: src/test/rgw/bucket_notification/test_bn.py [hunk 9]
```diff
@@ -1867,6 +1870,114 @@ def test_ps_s3_lifecycle_on_master():
...
+            endpoint_url='http://'+conn.host+':'+str(conn.port),
+            aws_access_key_id=conn.aws_access_key_id,
+            aws_secret_access_key=conn.aws_secret_access_key)
+    response = client.put_bucket_lifecycle_configuration(Bucket=bucket_name, 
+            LifecycleConfiguration={'Rules': [
+                {
+                    'ID': 'abort1',
+                    'Filter': {'Prefix': obj_prefix},
+                    'Status': 'Enabled',
+                    'AbortIncompleteMultipartUpload': {'DaysAfterInitiation': "1"},
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/test_bn.py:1942]:
```suggestion
                    'AbortIncompleteMultipartUpload': {'DaysAfterInitiation': 1},
```

File: src/rgw/rgw_lc.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (yuvalif) [reply src/rgw/rgw_lc.cc:831]: agree that it does not change anything with the compiled code.
however, it helps the readers of the code, when they try to track state in the code.

File: src/rgw/rgw_lc.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (yuvalif) [reply src/rgw/rgw_lc.cc:848]: ok. will fix later

File: src/rgw/rgw_lc.cc (hunk 2 above)
Comment (yuvalif) [reply src/rgw/rgw_lc.cc:863]: silent failures could be an issue. IMO, we should log

**Comment by yuvalif** [comment]
> > > i guess that transition could only be tested manually, but can you please add MPU abort test to `src/test/rgw/bucket_notification/test_bn.py`?
> > > see as an example: `test_ps_s3_lifecycle_on_master()`
> > 
> > 
> > will give it a shot (oh my, it uses nose...)
> 
> after experimenting with test_bn.py; it's clear that lifecycle transition, in particular, will be more difficult to add support for. checking into whether we have existing automation for transition setup. meanwhile, abortincompletemultipartupload has only a daysafterinitiation parameter--so we have to be using rgw_lc_debug_interval; is that the case right now?
> 
> In any case, I have a draft change to test_bn.py that should trigger lifecycle abort, presuming our radosgw is running with the rgw_lc_debug_interval that the lifecycle tests in s3-tests rely on, which seems likely. I'm not sure if boto2 (which this file currently uses) or boto3 (which it should use) might object to our eventName extension. I did find an apparent bug in the formatting that name, sending that too.

test passed when running locally (with the fix below)

File: src/rgw/rgw_lc.cc (hunk 4 above)
Comment (yuvalif) [reply src/rgw/rgw_lc.cc:1322]: they are all SAL objects, not "rgw types".
for example, in line 1317 you define an `rgw::sal::Bucket*` but `oc.bucket` is also `rgw::sal::Bucket*`.
what is the difference between these objects and why do you need to allocate the local bucket in line 1322?

same question regarding `user` and `obj`.

**Comment by mattbenjamin** [comment]
@yuvalif could you repost your bugfix?

**Comment by mattbenjamin** [comment]
@soumyakoduri could you re-review?  I don't see your comments after reopening this

File: src/test/rgw/bucket_notification/test_bn.py (hunk 9 above)
Comment (mattbenjamin) [reply src/test/rgw/bucket_notification/test_bn.py]: oh, ok

File: src/rgw/rgw_lc.cc (hunk 4 above)
Comment (mattbenjamin) [reply src/rgw/rgw_lc.cc:1329]: There were no sal types in lc_op_ctx until recently.  But additionally, creating those sal objects happens also in remove_expired_obj(), which Daniel reviewed over a year ago, so I followed the same pattern I'd used before.  It's possible this was an oversight, but it also changes more code than is being added, which I don't really like.  There is no SAL user in lc_op_ctx.

Comment (mattbenjamin) [reply src/rgw/rgw_lc.cc:1329]: I'll try revising both code paths, but we'll need to test more

**Comment by yuvalif** [comment]
@mattbenjamin, build fails with a compilation error

File: src/rgw/rgw_lc.cc (hunk 4 above)
Comment (dang) [reply src/rgw/rgw_lc.cc:1329]: So, looking at remove_expired_obj(), it looks like it was a mistake by me.  I think we can just use the versions on oc, and not make copies of them.

**Review by yuvalif — APPROVED** [review]

File: src/rgw/rgw_lc.cc [hunk 10]
```diff
@@ -1268,25 +1297,73 @@ class LCOpAction_Transition : public LCOpAction {
     /* If bucket is versioned, create delete_marker for current version
      */
     if (oc.bucket->versioned() && oc.o.is_current() && !oc.o.is_delete_marker()) {
-      ret = remove_expired_obj(oc.dpp, oc, false, rgw::notify::ObjectExpiration);
+      ret = remove_expired_obj(oc.dpp, oc, false, rgw::notify::ObjectTransitionCurrent);
       ldpp_dout(oc.dpp, 20) << "delete_tier_obj Object(key:" << oc.o.key << ") current & not delete_marker" << " versioned_epoch:  " << oc.o.versioned_epoch << "flags: " << oc.o.flags << dendl;
     } else {
```
Comment (soumyakoduri) [inline src/rgw/rgw_lc.cc:1302]: `else` case  here refers to both non-current objects and non-versioned objects. So it may be best to add another case here. 

if (!oc.bucket.versioned())  {
   ret = remove_expired_obj(oc.dpp, oc, true, rgw::notify::ObjectTransition);
} else if (oc.o.is_current() && !oc.o.is_delete_marker()) {
   ret = remove_expired_obj(oc.dpp, oc, false, rgw::notify::ObjectTransitionCurrent);
} else {
   ret = remove_expired_obj(oc.dpp, oc, true, rgw::notify::ObjectTransitionNoncurrent);
}

Comment (mattbenjamin) [reply src/rgw/rgw_lc.cc:1302]: thanks, let me revisit that

Comment (mattbenjamin) [reply src/rgw/rgw_lc.cc:1302]: I think the notification should still be ObjectTransitionCurrent?  that's how expiration works...

**Comment by cbodley** [comment]
@mattbenjamin it looks like there's at least 1 unresolved review comment from Soumya

**Comment by yuvalif** [comment]
@mattbenjamin ping

**Comment by mattbenjamin** [comment]
@yuvalif @soumyakoduri updated, at last (sorry)

File: src/rgw/rgw_lc.cc [hunk 11]
```diff
@@ -828,15 +824,47 @@ int RGWLC::handle_multipart_expiration(rgw::sal::Bucket* target,
...
+      ret = notify->publish_reserve(this, nullptr);
+      if (ret < 0) {
+				ldpp_dout(wk->get_lc(), 0)
+					<< "ERROR: reserving persistent notification for abort_multipart_upload, ret=" << ret
+					<< ", thread:" << wq->thr_name()
+					<< ", meta:" << obj.key
+					<< dendl;
+      }
+
+      ret = mpu->abort(this, cct, null_yield);
```
Comment (yuvalif) [inline src/rgw/rgw_lc.cc:855]: i understand that the operation should succeed even if the notification failed?

File: src/rgw/rgw_lc.cc [hunk 12]
```diff
@@ -828,15 +824,47 @@ int RGWLC::handle_multipart_expiration(rgw::sal::Bucket* target,
...
       if (ret == 0) {
-        if (perfcounter) {
+
+        ret = notify->publish_commit(
```
Comment (yuvalif) [inline src/rgw/rgw_lc.cc:858]: if `publish_reserved()` failed, the commit could fail as well, we can avoid calling it here

File: src/rgw/rgw_lc.cc (hunk 11 above)
Comment (mattbenjamin) [reply src/rgw/rgw_lc.cc:855]: yes, I think so

Comment (mattbenjamin) [reply src/rgw/rgw_lc.cc:855]: well, it looks like I had the opposite intuition in the original PR adding expire notificaitons.  It's definitely not acceptable to be inconsistent.  I think this really should be a policy decision--I can see a reasonable site policy preferring to reduce data by default, and also one that would prefer notifications to be reliable at the risk of data growth

File: src/rgw/rgw_lc.cc (hunk 12 above)
Comment (mattbenjamin) [reply src/rgw/rgw_lc.cc:858]: amended, as above

File: src/rgw/rgw_lc.cc (hunk 11 above)
Comment (yuvalif) [reply src/rgw/rgw_lc.cc:856]: agree we must be consistent here.
as for the policy, if the users does not need elivery guarantee for the notification. they can configure `ack-level=none` for the topic, this would mean that the queue entry is cleared once the notification is sent (not waiting for an ack), which means that the queue does get full even if the broker is down. in such a case `publish_reserve()` would fail only if there is a RADOS issue.

Comment (mattbenjamin) [reply src/rgw/rgw_lc.cc:856]: does rgw similarly refuse to perform an i/o request altogether if publish_reserve() fails?

Comment (yuvalif) [reply src/rgw/rgw_lc.cc:856]: yes, when `publish_reserve()` fails with queue full indication, we reply "BUSY" to the client

Comment (mattbenjamin) [reply src/rgw/rgw_lc.cc:856]: ok

**Comment by mattbenjamin** [comment]
https://pulpito.ceph.com/mbenjamin-2023-10-11_11:10:19-rgw-wip-notify-lctrans-distro-default-smithi/

**Comment by mattbenjamin** [comment]
https://pulpito.ceph.com/mbenjamin-2023-10-26_00:15:55-rgw-wip-notify-lctrans-distro-default-smithi/

**Comment by yuvalif** [comment]
> https://pulpito.ceph.com/mbenjamin-2023-10-26_00:15:55-rgw-wip-notify-lctrans-distro-default-smithi/

@AliMasarweh the test above is failing with:
```
2023-10-26T01:20:34.604 INFO:teuthology.orchestra.run.smithi029.stderr:======================================================================
2023-10-26T01:20:34.605 INFO:teuthology.orchestra.run.smithi029.stderr:FAIL: test persistent topic configurations with max_retries and retry_sleep_duration
2023-10-26T01:20:34.605 INFO:teuthology.orchestra.run.smithi029.stderr:----------------------------------------------------------------------
2023-10-26T01:20:34.605 INFO:teuthology.orchestra.run.smithi029.stderr:Traceback (most recent call last):
2023-10-26T01:20:34.605 INFO:teuthology.orchestra.run.smithi029.stderr:  File "/usr/lib/python3.6/site-packages/nose/case.py", line 197, in runTest
2023-10-26T01:20:34.605 INFO:teuthology.orchestra.run.smithi029.stderr:    self.test(*self.arg)
2023-10-26T01:20:34.606 INFO:teuthology.orchestra.run.smithi029.stderr:  File "/home/ubuntu/cephtest/ceph/src/test/rgw/bucket_notification/test_bn.py", line 3316, in test_ps_s3_persistent_topic_configs_max_retries
2023-10-26T01:20:34.606 INFO:teuthology.orchestra.run.smithi029.stderr:    ps_s3_persistent_topic_configs(persistency_time, config_dict)
2023-10-26T01:20:34.606 INFO:teuthology.orchestra.run.smithi029.stderr:  File "/home/ubuntu/cephtest/ceph/src/test/rgw/bucket_notification/test_bn.py", line 3253, in ps_s3_persistent_topic_configs
2023-10-26T01:20:34.606 INFO:teuthology.orchestra.run.smithi029.stderr:    assert_equal(parsed_result['Topic Stats']['Entries'], 0)
2023-10-26T01:20:34.606 INFO:teuthology.orchestra.run.smithi029.stderr:AssertionError: 4 != 0
```

could it be something you fixed the the past, and just requires a rebase?

**Comment by cbodley** [comment]
> @yuvalif @cbodley notwithstanding the above, this run looks like a pass to me; would you agree?

i agree, but i'd like to verify that the failure is unrelated. i don't recognize that from other runs

**Comment by mattbenjamin** [comment]
it seems like it if it is related, it doesn't have a fix in rgw_lc.cc?  but ok

**Comment by cbodley** [comment]
any way the new `test_ps_s3_lifecycle_abort_mpu_on_master()` could influence the topic stats in that later test?

**Comment by yuvalif** [comment]
> any way the new `test_ps_s3_lifecycle_abort_mpu_on_master()` could influence the topic stats in that later test?

should not impact the stats. each queue should have its own stats. IMO, could be an issue that was fixed on "main", and need a rebase

**Comment by AliMasarweh** [comment]
> > https://pulpito.ceph.com/mbenjamin-2023-10-26_00:15:55-rgw-wip-notify-lctrans-distro-default-smithi/
> 
> @AliMasarweh the test above is failing with:
> 
> ```
> 2023-10-26T01:20:34.604 INFO:teuthology.orchestra.run.smithi029.stderr:======================================================================
> 2023-10-26T01:20:34.605 INFO:teuthology.orchestra.run.smithi029.stderr:FAIL: test persistent topic configurations with max_retries and retry_sleep_duration
> 2023-10-26T01:20:34.605 INFO:teuthology.orchestra.run.smithi029.stderr:----------------------------------------------------------------------
> 2023-10-26T01:20:34.605 INFO:teuthology.orchestra.run.smithi029.stderr:Traceback (most recent call last):
> 2023-10-26T01:20:34.605 INFO:teuthology.orchestra.run.smithi029.stderr:  File "/usr/lib/python3.6/site-packages/nose/case.py", line 197, in runTest
> 2023-10-26T01:20:34.605 INFO:teuthology.orchestra.run.smithi029.stderr:    self.test(*self.arg)
> 2023-10-26T01:20:34.606 INFO:teuthology.orchestra.run.smithi029.stderr:  File "/home/ubuntu/cephtest/ceph/src/test/rgw/bucket_notification/test_bn.py", line 3316, in test_ps_s3_persistent_topic_configs_max_retries
> 2023-10-26T01:20:34.606 INFO:teuthology.orchestra.run.smithi029.stderr:    ps_s3_persistent_topic_configs(persistency_time, config_dict)
> 2023-10-26T01:20:34.606 INFO:teuthology.orchestra.run.smithi029.stderr:  File "/home/ubuntu/cephtest/ceph/src/test/rgw/bucket_notification/test_bn.py", line 3253, in ps_s3_persistent_topic_configs
> 2023-10-26T01:20:34.606 INFO:teuthology.orchestra.run.smithi029.stderr:    assert_equal(parsed_result['Topic Stats']['Entries'], 0)
> 2023-10-26T01:20:34.606 INFO:teuthology.orchestra.run.smithi029.stderr:AssertionError: 4 != 0
> ```
> 
> could it be something you fixed the the past, and just requires a rebase?

Yes I think I fixed this and haven't seen it for a while

**Comment by yuvalif** [comment]
after rebase, we get failure only with the new test of topic priority.
http://pulpito.front.sepia.ceph.com/yuvalif-2023-10-30_19:40:56-rgw:notifications-wip-yuval-notify-lctrans-distro-default-smithi/
```
2023-10-30T20:29:37.160 INFO:teuthology.orchestra.run.smithi059.stderr:======================================================================
2023-10-30T20:29:37.161 INFO:teuthology.orchestra.run.smithi059.stderr:ERROR: test s3 topic set/get/delete permissions
2023-10-30T20:29:37.161 INFO:teuthology.orchestra.run.smithi059.stderr:----------------------------------------------------------------------
2023-10-30T20:29:37.161 INFO:teuthology.orchestra.run.smithi059.stderr:Traceback (most recent call last):
2023-10-30T20:29:37.161 INFO:teuthology.orchestra.run.smithi059.stderr:  File "/home/ubuntu/cephtest/ceph/src/test/rgw/bucket_notification/virtualenv/lib64/python3.9/site-packages/nose/case.py", line 198, in runTest
2023-10-30T20:29:37.161 INFO:teuthology.orchestra.run.smithi059.stderr:    self.test(*self.arg)
2023-10-30T20:29:37.162 INFO:teuthology.orchestra.run.smithi059.stderr:  File "/home/ubuntu/cephtest/ceph/src/test/rgw/bucket_notification/test_bn.py", line 4379, in test_ps_s3_topic_permissions
2023-10-30T20:29:37.162 INFO:teuthology.orchestra.run.smithi059.stderr:    except ClientError as error:
2023-10-30T20:29:37.162 INFO:teuthology.orchestra.run.smithi059.stderr:NameError: name 'ClientError' is not defined
```

File: src/test/rgw/bucket_notification/test_bn.py [hunk 13]
```diff
@@ -7,18 +7,20 @@
 import socket
 import time
 import os
+import io
 import string
 import boto
-from botocore.exceptions import ClientError
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/test_bn.py:12]: this is causing a test failure. please put back this line

Comment (mattbenjamin) [reply src/test/rgw/bucket_notification/test_bn.py:12]: ok, will update shortly;  thanks for the review!

**Comment by mattbenjamin** [comment]
https://pulpito.ceph.com/mbenjamin-2023-11-01_23:31:48-rgw-wip-notify-lctrans-distro-default-smithi/

**Comment by mattbenjamin** [comment]
@cbodley @yuvalif I think this passed ^^

**Comment by mattbenjamin** [comment]
sorry about the submodule change, but given the scheduling overhead of teuthology, we shouldn't have to do re-runs if a few jobs don't start

**Comment by cbodley** [comment]
> sorry about the submodule change, but given the scheduling overhead of teuthology, we shouldn't have to do re-runs if a few jobs don't start

i know it's painful, but those failures could mask regressions. and the dead job there prevented the only rgw/notification job from running so we don't know it's fixed

**Comment by cbodley** [comment]
> The following tests FAILED:
	 31 - run-rbd-unit-tests-1.sh (Failed)

**Comment by cbodley** [comment]
jenkins test make check

**Comment by mattbenjamin** [comment]
so spurious failure?  mmm

**Comment by mattbenjamin** [comment]
https://pulpito.ceph.com/mbenjamin-2023-11-03_22:35:29-rgw-wip-notify-lctrans-distro-default-smithi/

**Comment by mattbenjamin** [comment]
https://pulpito.ceph.com/mbenjamin-2023-11-04_19:36:53-rgw-wip-notify-lctrans-distro-default-smithi/

**Comment by yuvalif** [comment]
> https://pulpito.ceph.com/mbenjamin-2023-11-04_19:36:53-rgw-wip-notify-lctrans-distro-default-smithi/

the notifications failure is a known kafka crash: https://tracker.ceph.com/issues/63314

**Comment by mattbenjamin** [comment]
thanks for reviewing the results, @yuvalif;  @cbodley this looks like a pass?
